pkg/manager/volumes/vol_compare_utils.go

```diff
@@ -0,0 +1,385 @@
+		if len(a) != len(b) {
+			return len(a) < len(b)
+		}
+		return a < b
+	})
+
+	return pods, nil
+}
+
+func (u *volCompareUtils) getStsOfComponent(cluster v1alpha1.Cluster, mt v1alpha1.MemberType) (*appsv1.StatefulSet, error) {
+	ns := cluster.GetNamespace()
+	stsName := controller.MemberName(cluster.GetName(), mt)
+
+	sts, err := u.deps.StatefulSetLister.StatefulSets(ns).Get(stsName)
+	if err != nil {
+		return nil, fmt.Errorf("get sts %s/%s failed: %w", ns, stsName, err)
+	}
+
+	return sts, nil
+}
+
+func (u *volCompareUtils) IsStatefulSetSynced(ctx *componentVolumeContext, sts *appsv1.StatefulSet, errOnNewVol bool) (bool, error) {
+	var newVolErr error = nil
+	for _, volTemplate := range sts.Spec.VolumeClaimTemplates {
+		volName := v1alpha1.StorageVolumeName(volTemplate.Name)
+		size := getStorageSize(volTemplate.Spec.Resources.Requests)
+		desired := getDesiredVolumeByName(ctx.desiredVolumes, volName)
+		if desired == nil {
+			// TODO: verify behaviour is correctly intended, this was previously being ignored. Now explicit error in
+			// pvc modifier..
+			errStr := fmt.Sprintf("volume %s in sts for cluster %s does not exist in desired volumes", volName, ctx.ComponentID())
+			if errOnNewVol {
+				newVolErr = errors.New(errStr)
+			} else {
+				klog.Warningf(errStr)
+			}
+			return false, newVolErr
+		}
+		if size.Cmp(desired.Size) != 0 {
+			return false, nil
+		}
+		scName := volTemplate.Spec.StorageClassName
+		if !isStorageClassMatched(desired.StorageClass, ignoreNil(scName)) {
+			return false, nil
+		}
+	}
+	// TODO: verify new behaviour is fine with pvc_modifier, previously number of volumes not checked.
+	if len(sts.Spec.VolumeClaimTemplates) != len(ctx.desiredVolumes) {
+		errStr := fmt.Sprintf("Number of volumes mismatch desired %d vs sts %d for cluster %s", len(ctx.desiredVolumes), len(sts.Spec.VolumeClaimTemplates), ctx.ComponentID())
+		if errOnNewVol {
+			newVolErr = errors.New(errStr)
+		} else {
+			klog.Warningf(errStr)
+		}
+		return false, newVolErr
+	}
+
+	return true, nil
+}
+
+func (u *volCompareUtils) IsPodSyncedForReplacement(ctx *componentVolumeContext, pod *corev1.Pod) (bool, error) {
+	// Does not check underlying StorageClass for modifications, only matches PVC specs for use by pvc replacer.
+	ns := pod.Namespace
+	podPvcCount := 0
+	for i := range pod.Spec.Volumes {
+		vol := &pod.Spec.Volumes[i]
+		pvc, err := u.getPVC(ns, vol)
+		if err != nil {
+			return false, err
+		}
+		if pvc == nil {
+			continue // Not a PVC don't compare.
+		}
+		podPvcCount++
+		desired := getDesiredVolumeByName(ctx.desiredVolumes, v1alpha1.StorageVolumeName(vol.Name))
+		if desired == nil {
+			// Extra un-desired volume in pod, needs to be removed.
+			return false, nil
+		}
+		desiredScName := desired.GetStorageClassName()
+		// If desired sc name is unspecified , any sc used by pvc is okay. (Default storage class).
+		if desiredScName != "" && ignoreNil(pvc.Spec.StorageClassName) != desiredScName {
```

Comment:
Should we also return `true` if desiredScName is set to empty string? If let's say a sc name is changed from a non-empty sc name to the empty string, I think user's intention is to switch from the original sc to the default sc, right? And maybe we should propagate the sc name to pvc?

Reply:
If desiredScName is blank, the pvc storage class will still not be blank but use whatever is the default storage class configured in the k8s cluster. It is a little tricky to find out which is default and use it to compare because the default might change later and trigger a change in disks with operator.

Existing pvc modifier also uses this behaviour (ignore changes if desired sc is blank), e.g.:
https://github.com/pingcap/tidb-operator/blob/899409c2f2858b6aa8e2f9a46db4607e32ee6fc4/pkg/manager/volumes/phase.go#L200
https://github.com/pingcap/tidb-operator/blob/899409c2f2858b6aa8e2f9a46db4607e32ee6fc4/pkg/manager/volumes/pod_vol_modifier.go#L403

So I think safer behavior is to not change storage class if it is not explicitly specified.